pkg/clustermesh/endpointslicesync/service_informer.go

```diff
@@ -191,17 +194,36 @@ type meshServiceLister struct {
 	namespace string
 }
 
+// List returns the matrix of all the local services and all the remote clusters.
+// This is not similar to what does the Get method. For instance, List may returns
+// services that could not be found on a call to the Get method.
+// By doing that we can ensure that the controller reconciliation is called
+// on every possible remote services especially the one that are deleted while
+// we still have some EndpointSlices locally (which won't be cleared by the
+// OwnerReference mechanism since our actual local service is not deleted).
 func (l meshServiceLister) List(selector labels.Selector) ([]*v1.Service, error) {
 	reqs, _ := selector.Requirements()
 	if !selector.Empty() {
 		return nil, fmt.Errorf("meshServiceInformer only supports listing everything as requirements: %s", reqs)
 	}
 
-	clusterSvcs := l.informer.globalServiceCache.GetServices(l.namespace)
-	svcs := make([]*v1.Service, 0, len(clusterSvcs))
-	for _, clusterSvc := range clusterSvcs {
-		if svc, err := l.informer.clusterSvcToSvc(clusterSvc, false); err == nil {
-			svcs = append(svcs, svc)
+	svcIter := l.informer.serviceStore.IterKeys()
+	clusters := l.informer.meshNodeInformer.ListClusters()
+	var svcs []*v1.Service
+	for svcIter.Next() {
+		if svcIter.Key().Namespace != l.namespace {
+			continue
+		}
```

Comment:
It could make sense to register a namespace indexer to the store and use `ByIndex` here, to avoid having to iterate over all services.

Reply:
Should I add that to `ServiceResource` in `pkg/k8s/resource_ctors.go` similarly to the indexer in `CiliumSlimEndpointResource`?

Reply:
Yeah, sounds reasonable to me.